database.cc

```diff
@@ -1791,16 +1791,16 @@ future<> database::apply_in_memory(const frozen_mutation& m, schema_ptr m_schema
 
     data_listeners().on_write(m_schema, m);
 
-  return cf.run_async([this, &m, m_schema = std::move(m_schema), h = std::move(h), &cf, timeout]() mutable {
+  return cf.run_async([this, &m, m_schema = std::move(m_schema), h = std::move(h), &cf, timeout] () mutable -> future<> {
     return cf.dirty_memory_region_group().run_when_memory_available([this, &m, m_schema = std::move(m_schema), h = std::move(h), &cf]() mutable {
         cf.apply(m, m_schema, std::move(h));
     }, timeout);
   });
 }
 
 future<> database::apply_in_memory(const mutation& m, column_family& cf, db::rp_handle&& h, db::timeout_clock::time_point timeout) {
-  return cf.run_async([this, &m, h = std::move(h), &cf, timeout]() mutable {
-    return cf.dirty_memory_region_group().run_when_memory_available([this, &m, &cf, h = std::move(h)]() mutable {
+  return cf.run_async([this, &m, h = std::move(h), &cf, timeout]() mutable -> future<> {
+    return cf.dirty_memory_region_group().run_when_memory_available([this, &m, &cf, h = std::move(h)] () mutable {
```

Comment:
This is a hot function, please check performance with perf-simple-query --smp 1 -c1 --operations-per-shard 1000000 --task-quota-ms 10. I'd like to see if allocs per op or instructions per op increase.

Reply:
(and --write!)

Reply:
Is it ok to run the test inside dbuild?

Reply:
The helper run_async does nothing but runs the code with _async_gate, perhaps we can drop the run_async completely and open code the with_gate.

Reply:
> Is it ok to run the test inside dbuild?

Yes, you may need to set some sysctl to allow perf_event_open to work in a container.

Reply:
> The helper run_async does nothing but runs the code with _async_gate, perhaps we can drop the run_async completely and open code the with_gate.

Yes, that's fine too.

Reply:
Done. I dropped run_async completely.